### ISSUE TITLE
Standardize #include

### ISSUE THREAD

**Issue by stefnotch** [bevyengine/naga_oil#83]
I've noticed that quite a few projects have started using naga oil. 🎉  This has lead to reasonable feature requests in WGSL language servers and WGSL tools:

- https://github.com/PolyMeilex/vscode-wgsl/issues/25 - wgsl-analyzer will need to re-implement the `#include` algorithm
- https://github.com/wgsl-analyzer/wgsl-analyzer/issues/88 - straightforward integration of naga_oil
- https://github.com/ScanMountGoat/wgsl_to_wgpu/issues/40 - naga_oil currently doesn't provide the required type info

Would it be possible to fully document how the `#include` algorithm works? And, potentially fix any rough edges that it might have at the moment?

I would be willing to help with this. Is there a preferred place for communication? :)

**Comment by stefnotch** [comment]
As for the `#include` algorithm and syntax:

Ideally, we'd write up a specification, and define a WGSL extension for this. This it to avoid having include statements get interpreted differently by different tools.

```wgsl
// Tells language servers which extension is going to be used. Theoretically, someone else could define a different wgsl importing extension, with a different name.
enable wgsl_imports;

 // Now this an unambiguous meaning
#import my_module;
```

See also https://github.com/gpuweb/gpuweb/issues/4576#issuecomment-2057893527 for how to not do it. (me complaining about GLSL tools)


Questions that aren't immediately clear to me from the readme:

- Where is a valid location for a `#import`? Only at the top of a file? Or inside functions?

- What would happen if there's any code before an `#import`. It shouldn't leak into the import, should it? 
  - e.g. `alias vec2f = u32;` is probably valid WGSL code, and can very much screw up code that comes after it.

- What exactly can be imported?
  - Structs: Yes
  - Functions: Yes 
  - [Extensions](https://www.w3.org/TR/WGSL/#extensions): ??
  - [Global diagnostic filters](https://www.w3.org/TR/WGSL/#global-diagnostic-directive): ???
  - Type aliases: ?
  - [Const declarations, override declarations](https://www.w3.org/TR/WGSL/#value-decls): Yes
  - [Var declarations](https://www.w3.org/TR/WGSL/#var-decls): Probably yes 
    - Importing a `var<private> foo: f32;` looks odd
    - Bindings might not compose nicely, because they have a fixed index `@group(0) @binding(2) var<uniform> param: Params;`


- How does the generation of the final shader work? I assume a fair bit of of the following is going on
  - Parsing the shaders
  - Tracking the used types
  - Name mangling

**Comment by stefnotch** [comment]
First of all, thank you so much for taking your time to respond to all of this!

> i'm completely open to adding/exposing features to aid integration directly. i'm not very keen on writing a formal spec, but happy to describe the current implementation (and potentially change the implementation if a spec does evolve).

Very understandable. I'm willing to do the work of writing up a specification.

> briefly:
> 
> > * What is a valid location for a `#import`?
> 
> anywhere is valid provided the import begins on its own line, but they are explicitly not scoped. imports are collected by the preprocessor and apply for the whole file, regardless of the "scope" of the import, whether it is declared before/after the usage, etc.

Alrighty. Would it be reasonable if the specification said something along the lines "top of a file only", and `naga_oil` simply ends up implementing a superset of the specification?
The intention there would be to not add tricky requirements for WGSL parsers that parse everything in one go.



> > ```
> > * Name mangling: Should have a documented algorithm
> > ```
> 
> we can expose this (we discussed that in another issue i think), but the current scheme is `{item}{pre_marker}{b64(module)}{post_marker}`, where `pre_marker` is a magic string that is otherwise disallowed (`X_NAGA_OIL_MOD_X` iirc), `post_marker` is `X`, `b64(module)` is the base64 encoding of the module's import path, and `item` is the undecorated identifier being imported.
> 
> there are some non-obvious criteria for this to produce valid identifier names in each target language (e.g. double underscores are not allowed in wgsl, naga doesn't handle items ending in digits or underscores well, putting the item first potentially helps with cache lookups, etc), and obviously if you make a proper spec the pre-marker would need to change.
> 

I think there's a name mangling scheme where one doesn't need a disallowed string. For example, using the ``{item}{pre_marker}{b64(module)}{post_marker}` scheme, we could always decode it by going backwards
1. Take the post marker
2. Take the b64 module. The pre_marker uses characters that don't appear in base64, so that's nice.
3. Take the pre_marker
4. Whatever is left is definitely the `{item}`

Now if `{item}` happens to include the `{pre_marker}` string, then that's not a problem. It's just a funnily named item, but doesn't cause any mangling or unmangling issues.

> 
> i'm normally on the bevy discord in rendering-dev (you might need to ping me though), and other people there are more or less familiar with naga oil as well.

Sweet, I just joined that Discord :) Looking forward to seeing where this goes, hopefully it'll end up being rather useful for the lovely bevy engie.

**Comment by robtfm** [comment]
The problem if you don’t have a magic marker is that users can make a raw item that happens to parse as a decorated item. I don’t see any way to avoid that without denying (part of) the marker. 

This could probably be managed with ast context, but I specifically tried to avoid that (naga doesn’t facilitate it currently, and it raises the bar for other language support significantly). 

> Would it be reasonable if the specification said something along the lines "top of a file only", and `naga_oil` simply ends up implementing a superset of the specification?

Sounds very reasonable to me.

**Comment by atlv24** [comment]
are imports preprocessor aware? can you conditionally import a file? do imported files take on the same defines as the first file?

**Comment by robtfm** [comment]
> are imports preprocessor aware? 

Yes - see eg the import of VertexOutput in bevy’s pbr_fragment.wgsl

> can you conditionally import a file? 

Yes, as above. Naga_oil also tree-shakes though, so you often don’t need to. 

> do imported files take on the same defines as the first file?

Yes, there is one set of defines applied through the whole module build, otherwise chaos would ensue.

**Comment by stefnotch** [comment]
@robtfm If we mangle every identifier, then I think we do not need a magic marker. This does mean mangling everything in the main module as well. This makes specifying [the entry point for a WGSL shader](https://docs.rs/wgpu/0.19.4/wgpu/struct.VertexState.html#structfield.entry_point) a bit more annoying, because a user could technically write something that we have to mangle

```rust
@compute
fn main_X_NAGA_OIL_MOD_X() {
}
```

I'll properly write it up later, and I'll also look at what Rust and Swift are doing for their name mangling, just so that I don't miss any essential details.

There's also an alternative mangling scheme, which does "escaping" instead. e.g. All identifiers in the main module are valid as is, except if we encounter something that could have a special meaning, then we escape it. Including escaping an escape sequence.
(Similar to how one escapes `"quotes \" and backslashes \\ in strings"`.)

**Comment by robtfm** [comment]
escaping won't work with the current status quo, because the modules need to be compiled by naga which enforces wgsl identifier rules (UAX31) - anything we can use, users can use too, unless we choose to deny something to users that's otherwise valid. so practically, escaping would have to be the same as using a magic marker and denying it to users.

mangling everything would work i think. we would also be able to undo that mangling on the top-level names (at least the entry points) once we've built the all pieces to naga, to avoid the ux issue. might be quite a lot slower though..

**Comment by stefnotch** [comment]
Counterintuitively, there is an escaping scheme which works under that constraint. (Very convenient for us)

We pick some escape character, like `A` (we'll pick a better one, don't worry). And then, we want to
1. Whenever the user tries to write something special, we escape it
2. Afterwards add the whole `{pre_marker}{b64(module)}{post_marker}` at the end of imported stuff

To decode it, we first look at the stuff at the end. Afterwards, we undo the escaping to get the original identifier.

As in, whenever the user writes
- `cat`: remains unchanged `cat`
- `main_X_NAGA_OIL_MOD_X`: gets escaped to `main_X_NAAGA_OIL_MOD_X`
- `AAcatA` gets escaped to `AAAAcatAA`
- `dog` from another module first gets escaped, and then we append the stuff `dog_X_NAGA_OIL_MOD...`
...

**Comment by stefnotch** [comment]
@robtfm I've been thinking about `#import`s for a while, and I have one major question: How does a language server find the relevant files?

Essentially, when a language server sees `#import foo`, then it should not have to search through all wgsl files in all folders to find the correct one to import. Instead, it should 
- either follow some easy-to-check rule like "look for a file with the name foo.wgsl in a specific folder"
- or use an import map, which is a file where all shaders are statically defined
- or something else?

**Comment by robtfm** [comment]
There’s an issue about making a cli which links to a simple one I knocked together. That shows how I would expect the resolution to work, in a lsp context you presumably have a known project root to search from, and additional folders could be configured from plugin settings (at least in code, I guess other containers would have a similar mechanism)

**Comment by robtfm** [comment]
Given the way resolution works there’s no way to avoid searching all files, but I guess you could generate a static map upfront and track changes if that’s too slow for building dynamically.

**Comment by stefnotch** [comment]
@robtfm Could we change how resolution will work in the future, to avoid searching all files?
I think the semantics of `#define_import_path` are suboptimal in that regard.

**Comment by robtfm** [comment]
If you’re sure it’s too onerous for servers to support then ok. Someone did actually post a working lsp to the bevy discord yesterday though…
